lib/compiler.js

```diff
@@ -606,21 +607,51 @@ function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
   return (fixedFile === undefined ? resolvedPath : fixedFile);
 }
 
-function buildAllMixinInstructions(appRootDir, mixinDirs, options) {
+function buildAllMixinInstructions(appRootDir, mixinDirs, mixinSources, options,
+  modelInstructions) {
   var extensions = _.without(_.keys(require.extensions),
     _.keys(getExcludedExtensions()));
+
   var files = options.mixins || [];
+  var mixins = loadMixins(appRootDir, mixinDirs, files, extensions, options);
+  if (mixins === undefined) return;
+
+  /* If `mixinDirs` and `mixinSources` have any directories in common,
+   * then remove the common directories from `mixinSources` */
+  mixinSources = _.difference(mixinSources, mixinDirs);
+  files = [];
+  var mixinsFromMixinSources = loadMixins(appRootDir, mixinSources, files,
+    extensions, options);
```

Comment:
Actually this is the place I find confusing regarding the `files` variable. If we go with the proposal in my previous comment, then this should become `loadMixins(appRootDir, mixinSources, null, extensions, options)`.

Or even better, drop the `files` (`sourceFiles`) parameter completely and implement the concatenation at the instructions level.